### ISSUE TITLE
AOT: add support for extending expressions in classes

### ISSUE THREAD

**Issue by schippie** [angular/angular#13436]
**I'm submitting a feature request**
```
[ ] bug report => search github for a similar issue or PR before submitting
[X ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Currently the AOT doesn't allow any functions to be within an annotation though this same rule applies to classes it seems. Meaning the following fully compatible typescript 1.6+ code, copied from the typescript What's new section (https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#extending-expressions), isn't allowed by the AOT:

```javascript
class ThingA {
    getGreeting() { return "Hello from A"; }
}

interface GreeterConstructor {
    new (): ThingA;
}

function getGreeterBase(): GreeterConstructor {
    return ThingA;
}

class Test extends getGreeterBase() {
    sayHello() {
        console.log(this.getGreeting());
    }
}
```

The area were this has manifested for me is in the use of immutablejs'es `Record` class (https://facebook.github.io/immutable-js/docs/#/Record):

```javascript
export class ModelA extends Record({name: ''}) {
     name: string
}
```

Which is something the AOT really doesn't like and throws an error on. 

**Expected behavior**
I'm not sure if the above is by design (though based on the response I got on gitter it would seem so) and due to this fact it's a feature request and not a bug report.  I would just like to know if it's by design or if there is a chance it will be altered, as it will influence the decisions I have to make going forwards when it comes to Ngrx/Store + ImmutableJS etc. 

**What is the motivation / use case for changing the behavior?**
The main reason I would like this changed is that it would bring the AOT in line with the regular Typescript compiler. While I understand functions directly inside of an annotation isn't something you would want I find it weird that the AOT parts here from the Typescript standard (which isn't clearly documented anywhere as far as I can tell).

* **Angular version:** 2.0.X
2.3.0
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
Typescript 2.0.10

* **Node (for AoT issues):** `node --version` =   
4.2.2

**Comment by robwormald** [comment]
@schippie I presume you're using this kind of thing to populate initial ngrx state in an annotation, vs using these things as injectables directly, right?

In theory, doing something like 

```ts
export function getRecord(){ return new ModelA() }
{ provide: SOME_TOKEN, useFactory: getRecord }
```
should solve the annotation issue, but I believe we need to fix ngrx to accept that. 

Still a useful feature here nonetheless.

**Comment by schippie** [comment]
@robwormald  The only place I would be using the code within the annotation would be within the creation of the store. Here I populate the store with initial data, though even there I purely create an empty map for the player example below. And I do not create a new instance of a `Player` class within the `NgModule` annotation. I do create a new instance for sub stores that are not a Collection of Records though. An example of this would be a User class of which only one exists within the application.

## model.ts ##
```javascript
import { Map, Record } from "immutable";

import { Util } from "../../util";

export class PlayerActions {
	static ADD = 'players.ADD';
	static UPDATE = 'players.UPDATE'
}

export interface PlayersState extends Map<number, Player> {}

export interface PlayerState {
    id: number
    userName: string
    firstName: string
    surName: string
    birthDate: Date
    gender: 'm' | 'f' | 'o'
    lastLoginDate: Date
    email: string
    avatarId: number
}

export class Player extends Record({
    id: 0,
    userName: '<<LOADING>>',
    firstName: '',
    surName: '',
    birthDate: new Date(),
    gender: 'o',
    lastLoginDate: new Date(),
    email: '',
    avatarId: 0
}) implements PlayerState {
    id: number;
    userName: string;
    firstName: string;
    surName: string;
    birthDate: Date;
    gender: 'm' | 'f' | 'o';
    lastLoginDate: Date;
    email: string;
    avatarId: number;
    
    /**
     * Function that returns the avatar image url. This is based on the specified
     * avatarId in combination with the width and height that get passed along to this function. 
     */
    getAvatarImage(width, height): string {
        return Util.getImageUrl(this.avatarId, width, height);      
    }
}

export const initialPlayersState: PlayersState = <PlayersState>Map();
```

## reducer.ts ##
```javascript
import { ActionReducer, Action } from "@ngrx/store";
import { PlayersState, Player, PlayerActions, initialPlayersState } from "./model";

export const players: ActionReducer<PlayersState> = (state: PlayersState = initialPlayersState, action: Action) => {
	switch(action.type) {
		case PlayerActions.ADD:
			return state.withMutations(map => {
				let player = new Player(action.payload);
				if(player.id<= 0) throw new Error("Their was no player id within the action.payload");
				map.set(player.id, player);		
			});
		case PlayerActions.UPDATE: 
			return state.update(action.payload.id, player => {
				if(player == null) throw new ReferenceError("Player with id: " + action.payload.mtId + " does not exist");
				return <Player>player.merge(player, action.payload);
			});
		default:
			return state;
	}
};
```

## store.ts ##
```javascript
...

export interface ApplicationState {
    ...
    players: PlayersState
}

//the application store
export const applicationStore = StoreModule.provideStore(combineReducers({
       ...
	players
}));
```

## module.ts (core feature module) ## 
```javascript
...

@NgModule({
	imports: [CommonModule, applicationStore],
	providers: [
		...RESOURCE_PROVIDERS,
		...UTILS_PROVIDERS,
		...ADAPTER_PROVIDERS
	]
})
export class CoreModule {}
```

**Comment by schippie** [comment]
@robwormald  Actually I decided to downgrade to version 2.2.3 and the errors related to the Record class have disappeared. Would this count as a regression then?

**Comment by tbosch** [comment]
@schippie your example seems strange, as you add the `applicationStore` into `@NgModule.imports`, where only module references are allowed. If this worked in 2.2.3 then this was a missing error message.

There are 2 cases here:

1) AoT should not throw if the above code is not referenced (even not transitively) from an Angular decorator like `@NgModule`, ... If this is the case, then this is a bug.
2) If you are referencing the above code in an Angular decorator, this is supposed to throw (by design), unless you refer to it via a factory provider, e.g.`{provide: someToken, useFactory: someFunctionThatUsesTheCodeAbove}`.

Do you have a reproduction for case 2) in a repo so that we can have a look?

**Comment by schippie** [comment]
@tbosch Hey Tobias, thank you for replying as far as I'm aware I'm already using the AOT compatible way (which is indeed slightly different from the comment due to the changes I still had to make to get everything working correctly in version 2.2.3 for the project we are working on) I have created a sample repository of the minimal amount of code needed to recreate the problem here:

https://bitbucket.org/schippie/aot-13436/overview

If I am just doing something wrong please say so. But to me it already seems like that is the correct way to provide the store module. 

Yet the same error remains (in version 2.4.1):

```
Error: Error encountered resolving symbol values statically. Symbol reference expected (position 13:29 in the original .ts file), resolving symbol Player in G:/frontend/record-test/src/core/resource/player/model.ts
    at positionalError (G:\frontend\record-test\node_modules\@angular\compiler\bundles\compiler.umd.js:25874:37)
    at simplifyInContext (G:\frontend\record-test\node_modules\@angular\compiler\bundles\compiler.umd.js:25711:29)
    at StaticReflector.simplify (G:\frontend\record-test\node_modules\@angular\compiler\bundles\compiler.umd.js:25725:15)
    at StaticReflector.annotations (G:\frontend\record-test\node_modules\@angular\compiler\bundles\compiler.umd.js:25219:82)
    at NgModuleResolver.resolve (G:\frontend\record-test\node_modules\@angular\compiler\bundles\compiler.umd.js:17583:84)
    at CompileMetadataResolver.getNgModuleMetadata (G:\frontend\record-test\node_modules\@angular\compiler\bundles\compiler.umd.js:18091:62)
    at addNgModule (G:\frontend\record-test\node_modules\@angular\compiler\bundles\compiler.umd.js:24985:60)
    at G:\frontend\record-test\node_modules\@angular\compiler\bundles\compiler.umd.js:24996:16
    at Array.forEach (native)
    at _createNgModules (G:\frontend\record-test\node_modules\@angular\compiler\bundles\compiler.umd.js:24995:28)
```

**Comment by zahlprish** [comment]
@schippie I am having the same exact issue when utilizing Immutable.js's Record class, too. Unfortunately I did not make any attempts prior to 2.4.0, but looking at your example we are pretty much utilizing the same pattern. Have you had any luck with any workarounds or anything?

**Comment by schippie** [comment]
@zahlprish  Well the moment i started to use 2.2.3 the problems vanished for me and the application in production is working perfectly.

**Comment by tbosch** [comment]
@schippie You are doing this:
```
@NgModule({
	imports: [CommonModule, applicationStore],
	providers: [
		...RESOURCE_PROVIDERS
	]
})
export class CoreModule {}

...

const applicationStore = new StoreModule.provideStore(reducer);
...
```

In an `@NgModule`, you can only list other classes that are annotated with `@NgModule` in the `imports` field. What is your goal with this code?

**Comment by schippie** [comment]
@tbosch Ah let me try to change that then, and no real goal only to keep everything in one file that was related to the creation of the store. I'll change it and see if the error goes away when I add `StoreModule.provideStore(reducer)` directly within the imports.

**Comment by schippie** [comment]
@tbosch  Alright I decided to take it a step further, I have simplified the example greatly and have managed to reproduce it. The moment a record is imported or referenced in anyway or form it throws the above error. When no record extending class is referenced no error occurs:

Test2:  https://bitbucket.org/schippie/aot-13436/branch/test2 (no store anymore and only one simple model.ts file containing the player model) **error**
Test3: https://bitbucket.org/schippie/aot-13436/branch/test3 ( use files array in ts-config instead of exclude array eliminating the ts-config directly referencing the model.ts file + changes made for Test 2 ) **error**
Test4: https://bitbucket.org/schippie/aot-13436/branch/test4  ( Test 2 changes + Test 3 changes + remove all references / import statements to model.ts ) **compiles**

**Comment by zahlprish** [comment]
@schippie any luck with this? Do we know if it if a bug or working as expected?

**Comment by schippie** [comment]
@zahlprish Not sure yet if it's a bug or working as expected, hopefully we will learn more the comming days / early next week.

**Comment by simplifia** [comment]
Hello there,

We've encoutered the same thing, and in order to use the Immutables Records, we started using this as a workaround :
https://github.com/rangle/typed-immutable-record

If it can helps :)

**Comment by schippie** [comment]
@simplifia I believe there is another workaround everyone can do as well which involves creating a record like (unverified code just writing it here hehe):

```javascript

export interface testModel {
   foo: string;
}

export const test = new Record({foo: "test"});

export class Dummy extends test implements testModel {
     foo: string;
}
```

But honestly those are just solutions around the problem that manifest due to the fact that extending expressions are not working as expected.

**Comment by simplifia** [comment]
We've tried what you proposed :

`import { Record } from 'immutable';

export interface TestModel {
  foo: string;
}

export const test: Record.Class = Record({ foo: 'test' });

export class Dummy extends test implements TestModel {
  foo: string;
}`

then 

`let foobar = new Dummy();
console.log('foobar', foobar);`

Output

`Error: Error encountered resolving symbol values statically. Calling function 'Record', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, resolving symbol test in /home/homer/DEV/NG2/simpli-ng2-framework/node_modules/@simplifia/ng2-framework/pagination/pagination.model.ts, resolving symbol Dummy in /home/homer/DEV/NG2/simpli-ng2-framework/node_modules/@simplifia/ng2-framework/pagination/pagination.model.ts`

**Comment by zahlprish** [comment]
I've still been avoiding a workaround to this point, but thinking about trying the one @simplifia provided.

**Comment by zahlprish** [comment]
Would it be possible that Angular itself is working as expected, but the update exposed a bug in the angular-cli?

**Comment by schippie** [comment]
@zahlprish I don't use angular-cli personally the problem is within the angular/compiler-cli

**Comment by zahlprish** [comment]
@schippie Ah, okay. Well that's at least good to know. I'll be the first to admit I am fairly new to the Angular/Javascript community. Just trying to brain dump any possible ideas (and keep this issue alive).

**Comment by nadavsinai** [comment]
@tbosch We also have the same immutable record /ngrx store interplay and our app is incapable of AoT because of this issue, I will be happy to help in any way I can. thanks

**Comment by dakolech** [comment]
I have similar issue when compiling project with aot flag:

> ERROR in Error encountered resolving symbol values statically. Calling function 'Record', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function, resolving symbol categoryRecord in /Users/dako/briisk/briisk-blog/briisk-blog/src/common/models/category.model.ts, resolving symbol Category in /Users/dako/briisk/briisk-blog/briisk-blog/src/common/models/category.model.ts

```
import { Record } from 'immutable';

export const categoryRecord = Record({
  name: undefined,
  id: undefined,
  updatedAt: undefined,
  createdAt: undefined
});

export class Category extends categoryRecord {
  public name: string;
  public id: number;
  public updatedAt: string;
  public createdAt: string;

  static fromJSON(jsonObject: any = {}): Category {
    jsonObject.updatedAt = new Date(Date.parse(jsonObject.updatedAt)).toLocaleString();
    jsonObject.createdAt = new Date(Date.parse(jsonObject.createdAt)).toLocaleString();
    return new Category(jsonObject);
  }

  constructor(
    name: string|any = '',
    id?: number,
    updatedAt?: string,
    createdAt?: string
  ) {
    if (typeof name === 'object') {
      super(name);
    } else if (typeof name === 'string') {
      super({ name, id, updatedAt, createdAt });
    }
  }
}
```

angular-cli: 1.0.0-beta.25.5
angular: 2.4.4

**Comment by artaommahe** [comment]
Same issue with code for thirdparty lib that does not related to ng2 at all. It's out of ng app code scope and just imports to be execuded to register editor custom format.
```ts
const BlockEmbed: IBlotConstructor = Quill.import('blots/block/embed');

export class ImageBlockBlot extends BlockEmbed {
```
```
ERROR in Error encountered resolving symbol values statically.
Symbol reference expected (position 16:37 in the original .ts file),
resolving symbol ImageBlockBlot
```

**Comment by schippie** [comment]
I really hope there will be a status update on this in some way or form as it will decide if we have to refactor all our models or not.

**Comment by schippie** [comment]
@alexeagle Will the change made for #14154 / #14158  also fix the problem describe by me within this task?

**Comment by alexeagle** [comment]
I haven't followed this whole thread - the change in #14158 suppresses the error you had gotten from the AOT compiler when it encounters an expression following `extends` but it doesn't collect any metadata from it either. I could imagine a scenario that is still not supported, for example, if the supertype expression evaluates to an `@Component` that you need as your supertype, we won't gather the Inputs and Outputs from it and so the generated code won't behave like a subtype.

As long as your usage of extend expressions doesn't require the Angular compiler to understand anything about the supertype, this should work fine.

Could you verify this is fixed for you in the snapshot builds (eg. `npm install angular/core-builds`) ?
I'll close this for now assuming you just needed the error to be suppressed.

**Comment by zahlprish** [comment]
@schippie I just finished upgrading my Angular setup. Went from 2.4.0 to 2.4.6 - still had the same issue. The main difference was that I had explicitly flag my prod build to NOT use AoT, as it seemed to do it by default (ng build --prod --aot false). After that it built as normal (without AoT).

Saw the updates here and the fix that alexeagle provided (to suppress the error) in 4.0.0-beta.6, so I upgraded from 2.4.6 to 4.0.0-beta.6. Now everything appears to be build successfully with or without the AoT flag, though I can not confirm nor deny that I seen any improvements in the file size yet (which is  kind of what I was expecting).

Either way, suppressing the error message sounds like the correct "fix". The classes in which I am extending do not carry any metadata or annotations, so this works for me. 

Now I can try to see what I can do with the AoT compiler without making major code changes.

**Comment by schippie** [comment]
@zahlprish That's great to hear! That means we at least have an upgrade path again ourselves. I hadn't had the time to test it out myself yet, but if it's working for you then I'm convinced it will work for me as well. 

Now the rest of our team can wait for it to become available within the angular-cli (after angular 4.0.0 comes out of beta). And I can't wait to use it as soon as angular 4.0.0 is out as I use systemjs (don't have to wait for the angular-cli) 

@alexeagle Thank you very much for fixing the problems this saves our team a huge amount of work and allows us to continue using the tools we prefer.
